src/runtime-rs/crates/hypervisor/src/firecracker/inner_device.rs

```diff
@@ -0,0 +1,56 @@
+//Copyright (c) 2019-2022 Alibaba Cloud
+//Copyright (c) 2019-2022 Ant Group
+//Copyright (c) 2023 Nubificus Ltd
+//
+//SPDX-License-Identifier: Apache-2.0
+
+use super::FcInner;
+use crate::{DeviceType, VmmState};
+use anyhow::{anyhow, Context, Result};
+
+impl FcInner {
+    pub(crate) async fn add_device(&mut self, device: DeviceType) -> Result<DeviceType> {
+        let localdevice = device.clone();
+        if self.state == VmmState::NotReady {
+            info!(
+                sl!(),
+                "VMM not ready, queueing device {}",
+                localdevice.clone()
+            );
+
+            self.pending_devices.insert(0, localdevice.clone());
+
+            return Ok(localdevice);
+        }
+
+        info!(sl!(), "FcInner: Add Device {} ", localdevice.clone());
+
+        let _ = match localdevice.clone() {
+            DeviceType::Block(block) => self
+                .hotplug_block_device(block.config.path_on_host.as_str(), block.config.index)
+                .await
+                .context("add block device"),
+            DeviceType::Network(network) => self
+                .add_net_device(&network.config, network.device_id)
+                .await
+                .context("add net device"),
+            _ => Err(anyhow!("unhandled device: {:?}", localdevice)),
+        };
+
+        Ok(localdevice)
```

Comment:
Is all the device cloning necessary here?  `device` is cloned into `localdevice` right at the top so if the function always returns `device` then `localdevice` can safely be consumed in the course of function execution, right?

Reply:
ack, we have reworked the add_device function to capture more devices, even if this is needed quite early in the FC case.